src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Bicep.Core.Emit
+{
+    public sealed class FunctionPlacementValidatorVisitor : SyntaxVisitor
+    {
+        private enum VisitedElement
+        {
+            Module,
+            ModuleParams
+        }
+
+        private readonly SemanticModel semanticModel;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        private readonly Stack<(SyntaxBase syntax, Symbol? symbol)> visitedSymbolsStack = new();
+        private readonly Stack<VisitedElement> visitedElementsStack = new();
+
+        private class VisitorSymbolScope : IDisposable
```

Comment:
>IDisposable [](http://example.com/codeflow?start=43&length=11)

This is a really neat way of managing the stack :)

Reply:
thanks :)

Reply:
@majastrz - I extracted it into a separate class so it can be reused in other visitors if needed.